enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,363 @@
+only the details of the DNS name will be kept which matches with the Name field of the CR.
+
+
+#### Variation [optional]
+
+
+### API Extensions
+
+The validation of ``DNSName`` field in ``EgressFirewallDestination`` will be updated to accept wildcard DNS names as well.
+It will be updated from ``^([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.?$`` which accepts only regular DNS names to
+``^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.?$``.
+
+````go
+// EgressFirewallDestination is the endpoint that traffic is either allowed or denied to
+type EgressFirewallDestination struct {
+	// ..
+
+	// dnsName is the domain name to allow/deny traffic to. If this is set, cidrSelector must be unset.
+	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.?$
+	DNSName string `json:"dnsName,omitempty"`
+}
+````
+
+The following ``DNSName`` CRD will be added to the ``k8s.ovn.org`` api-group.
+
+````go
+// DNSName describes a DNS name used in a EgressFirewall rule.
+type DNSName struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the desired behavior of DNSName.
+	Spec DNSNameSpec `json:"spec"`
+	// Observed status of DNSName
+	// +optional
+	Status DNSNameStatus `json:"status,omitempty"`
+}
+
+// DNSNameSpec is a desired state description of DNSName.
+type DNSNameSpec struct {
+	// IsRegular denotes whether the resource is representing a regular DNS name.
+	// Cannot be false if IsWildcard is also false.
+	// +optional
+	IsRegular bool `json:"isregular,omitempty"`
+	// IsWildcard denotes whether the resource is representing a wildcar DNS name.
+	// Cannot be false if IsRegular is also false.
+	// +optional
+	IsWildcard bool `json:"iswildcard,omitempty"`
+}
+
+type DNSNameStatus struct {
+	// The list of matching DNS names and their corresponding IPs along with TTL and next
+	// time to DNS lookup.
+	Items []DNSNameStatusItem `json:"items"`
+}
+
+type DNSNameStatusItem struct {
+	DNSName string        `json:"dnsname"`
+	Info    []DNSNameInfo `json:"info"`
+}
+
+type DNSNameInfo struct {
+	IP             string `json:"ip"`
+	TTL            string `json:"ttl"`
+	NextLookupTime string `json:"nextlookuptime"`
+}
+````
+
+### Implementation Details/Notes/Constraints [optional]
+
+The implementation changes needed for the proposed enhancement are documented in this section for each of the components.
+
+#### Cluster DNS Operator
+
+Cluster DNS Operator will deploy CoreDNS with the ``egressfirewall`` plugin enabled by adding it to the corefile. As the
+plugin will watch and update the ``DNSName`` CRs in the ``k8s.ovn.org`` api-group, proper RBAC permissions will be needed
+to be added to the ``ClusterRole`` for CoreDNS.
+
+#### CoreDNS
+
+The new plugin ``egressfirewall`` will be added to CoreDNS. As the plugin will inspect the DNS lookup queries and response from
+other plugins, it needs to be added before the other plugins (namely ``forward`` plugin) in the ``plugin.cfg`` file.
```

Comment:
> Will the new plugin go before or after the cache plugin in `plugin.cfg`?

Would it be better to not add the new plugin before cache plugin? As cache plugin won't provide new IPs, we can reduce the time spent in inspecting responses from cache plugin.



> Will it be added to all server blocks, or just the main `.:5353` one in the Corefile?

Should it be added to all server blocks? I was thinking of adding it to only the main `.:5353` server block.

Reply:
> > Would it be better to not add the new plugin before cache plugin? As cache plugin won't provide new IPs, we can reduce the time spent in inspecting responses from cache plugin.
> 
> I'm not sure. If the egressfirewall plugin comes before the cache plugin, then the egressfirewall plugin's logic will run for cached responses. A cached response will be the same as the original response, except that its TTL is reduced by the time it has already been cached. So if I understand correctly, the egressfirewall plugin would update the DNSName CR with the new TTL and next-lookup-time values. That's the only functional difference I can think of between putting egressfirewall first and putting cache first.

As only the TTL values are reduced by the time it has been cached, will the next lookup time change? Also, the status of the DNSName CRs are updated only when the associated IPs change, otherwise it's not updated. Wouldn't it be better to avoid the inspection of the response from the cache plugin by the new plugin as it will not trigger any updates?



> > > Will it be added to all server blocks, or just the main `.:5353` one in the Corefile?
> > 
> > 
> > Should it be added to all server blocks? I was thinking of adding it to only the main `.:5353` server block.
> 
> I think you want it for all server blocks because you want EgressFirewall rules to apply consistently even for DNS names that are resolved by custom upstreams—or is there some reason this wouldn't be expected and desirable?

Ok, it makes sense. Will add it to all the server blocks.

Reply:
There is a potential issue with the `ocp_dnsnameresolver` plugin being placed below `cache` plugin. If a DNS name is looked up and then before the TTL of the corresponding IP addresses expire an EgressFirewall with a DNS rule (or a DNSNameResolver) is created for the same DNS name, then the subsequent DNS lookups, until the TTL expires, will be served by the `cache` plugin. As a result, for the time till the TTL expires, the DNSNameResolver object for the corresponding DNS name will not be updated with the IP addresses. This problem will be resolved if we put the `ocp_dnsnameresolver` plugin before the `cache` plugin.